dango/types/src/vesting.rs

```diff
@@ -0,0 +1,139 @@
+
+#[grug::derive(Serde)]
+pub struct InstantiateMsg {}
+
+#[grug::derive(Serde)]
+pub enum ExecuteMsg {
+    CreatePosition {
+        user: Addr,
+        schedule: Schedule<Option<Timestamp>>,
+    },
+    Claim {
+        idx: u64,
+    },
+}
+
+#[grug::derive(Serde, QueryRequest)]
+pub enum QueryMsg {
+    #[returns(ClaimablePosition)]
+    Position { idx: u64 },
+    #[returns(BTreeMap<u64, ClaimablePosition>)]
+    Positions {
+        start_after: Option<u64>,
+        limit: Option<u32>,
+    },
+    #[returns(BTreeMap<u64, ClaimablePosition>)]
+    PositionsByUser {
+        user: Addr,
+        start_after: Option<u64>,
+        limit: Option<u32>,
+    },
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Schedule<T = Timestamp> {
+    pub start_time: T,
+    pub cliff: Option<Duration>,
+    pub vesting: Option<Duration>,
+}
+
+impl Schedule<Option<Timestamp>> {
+    pub fn set_start_time(self, now: Timestamp) -> anyhow::Result<Schedule> {
+        let start_time = if let Some(start_time) = self.start_time {
+            ensure!(
+                start_time >= now,
+                "invalid start time: {:?}, now: {:?}",
+                start_time,
+                now
+            );
+            start_time
+        } else {
+            now
+        };
+
+        Ok(Schedule {
+            start_time,
+            cliff: self.cliff,
+            vesting: self.vesting,
+        })
+    }
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Position<C = Undefined> {
+    pub user: Addr,
+    pub schedule: Schedule,
+    pub amount: Coin,
+    pub claimed_amount: Uint128,
+    pub claimable_amount: C,
+}
+
+impl Position {
+    pub fn new(user: Addr, schedule: Schedule, amount: Coin) -> Self {
+        Self {
+            user,
+            schedule,
+            amount,
+            claimed_amount: Uint128::ZERO,
+            claimable_amount: Undefined::new(),
+        }
+    }
+
+    pub fn with_claimable_amount(self, now: Timestamp) -> Position<Uint128> {
+        let claimable_amount = self.compute_claimable_amount(now).unwrap_or_default();
+        Position {
+            user: self.user,
+            schedule: self.schedule,
+            amount: self.amount,
+            claimed_amount: self.claimed_amount,
+            claimable_amount,
+        }
+    }
+}
+
+impl<T> Position<T> {
+    pub fn compute_claimable_amount(&self, now: Timestamp) -> anyhow::Result<Uint128> {
+        ensure!(
+            now >= self.schedule.start_time,
+            "vesting has not started yet: start at: {:?}, now: {:?}",
+            self.schedule.start_time,
+            now
+        );
+
+        let mut time = self.schedule.start_time;
+
+        if let Some(cliff) = self.schedule.cliff {
+            time = time + cliff;
+
+            ensure!(
+                now >= time,
+                "nothing to claim during cliff phase: end: {:?}, now: {:?}",
+                time,
+                now
+            );
+        }
```

Comment:
I don't see why throwing error here. If claimable amount is zero then just return zero.

Reply:
It was just to have a dedicated error instead to have a zero amount error